google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/test_helpers/env/AbstractTestEnv.java

```diff
@@ -127,57 +176,160 @@ public String getSecondaryZone() {
-        getInstanceAdminClient().deleteAppProfile(getInstanceId(), appProfileId, true);
+      if (!appProfile.getId().startsWith(PREFIX)) {
+        continue;
+      }
+      boolean isNewerThanStale = appProfile.getId().compareTo(stalePrefix) > 0;
+      if (isNewerThanStale) {
+        continue;
+      }
+      try {
+        getInstanceAdminClient().deleteAppProfile(getInstanceId(), appProfile.getId(), true);
+      } catch (NotFoundException ignored) {
+
       }
     }
   }
 
-  private void cleanUpStaleClusters() {
-    String staleClusterId = TEST_CLUSTER_PREFIX + Instant.now().minus(1, ChronoUnit.DAYS);
-
+  /**
+   * Clean up clusters that were dynamically created in the default instance that have been
+   * orphaned.
+   *
+   * @param stalePrefix
+   */
+  private void cleanUpStaleClusters(String stalePrefix)
+      throws ExecutionException, InterruptedException {
     for (Cluster cluster : getInstanceAdminClient().listClusters(getInstanceId())) {
-      String clusterId = cluster.getId();
-      if (clusterId.startsWith(staleClusterId) && staleClusterId.compareTo(clusterId) > 0) {
-        getInstanceAdminClient().deleteCluster(getInstanceId(), clusterId);
+      if (!cluster.getId().startsWith(PREFIX)) {
+        continue;
+      }
+      boolean isNewerThanStale = cluster.getId().compareTo(stalePrefix) > 0;
+      if (isNewerThanStale) {
+        continue;
+      }
+
+      try {
+        deleteCluster(getTableAdminClient(), cluster.getId());
+      } catch (NotFoundException ignored) {
+
+      }
+    }
+  }
+
+  /**
+   * Clean up dynamically created (non-default) instances that have been orphaned.
+   *
+   * @param stalePrefix
+   */
+  private void cleanUpStaleInstances(String stalePrefix)
+      throws IOException, ExecutionException, InterruptedException {
+    for (Instance instance : getInstanceAdminClient().listInstances()) {
+      if (!instance.getId().startsWith(PREFIX)) {
+        continue;
+      }
+      boolean isNewerThanStale = instance.getId().compareTo(stalePrefix) > 0;
+      if (isNewerThanStale) {
+        continue;
+      }
+      try {
+        deleteInstance(instance.getId());
+      } catch (NotFoundException ignored) {
+
+      }
+    }
+  }
+
+  /** Delete an instance with all of its resources. */
+  private void deleteInstance(String instanceId)
+      throws IOException, ExecutionException, InterruptedException {
+    BigtableTableAdminSettings settings =
+        getTableAdminSettings().toBuilder().setInstanceId(instanceId).build();
+
+    // Delete all child resources (backups & clusters) that wont be automatically deleted
+    try (BigtableTableAdminClient tableAdmin = BigtableTableAdminClient.create(settings)) {
+      List<Cluster> clusters = getInstanceAdminClient().listClusters(instanceId);
+
+      boolean isFirstCluster = true;
+
+      for (Cluster cluster : clusters) {
+        deleteCluster(tableAdmin, cluster.getId());
+        if (!isFirstCluster) {
```

Comment:
why do we need this check?

Reply:
I added a comment, but the tldr version is that we can't delete the last remaining cluster of an instance, so I leave the very first one around for deleteInstance to delete